First iteration over transport draft (#1857)
quicwg/base-drafts: draft-ietf-quic-transport.md

the protocol to evolve without incurring a dependency on upgrades to
middleboxes.

> This document describes the core QUIC protocol, and is structured as follows:

Reviewer: I like two levels

> : Indicates that x is variable-length

Reviewer 2: This section on encoding and pictures could be moved much later in the document.  We don't really use this until we start describing the wire format, and if we had a section on notation conventions before we got into packet formats and the like, we could put this with the note on network byte order and varint encoding.

with small changes is undesirable.  At the same time, infrequent updates require
larger increments to limits if blocking is to be avoided.  Thus, larger updates
require a receiver to commit to larger resource commitments.  Thus there is a
trade-off between resource commitment and overhead when determining how large a
limit is advertised.

A receiver MAY use an autotuning mechanism to tune the frequency and amount that
it increases data limits based on a round-trip time estimate and the rate at
which the receiving application consumes data, similar to common TCP
implementations.

## Stream Limit Increment

As with flow control, this document leaves when and how many streams to make
available to a peer via MAX_STREAM_ID to implementations, but offers a few
considerations.  MAX_STREAM_ID frames constitute minimal overhead, while
withholding MAX_STREAM_ID frames can prevent the peer from using the available
parallelism.

Implementations will likely want to increase the maximum stream ID as
peer-initiated streams close.  A receiver MAY also advance the maximum stream ID
based on current activity, system conditions, and other environmental factors.

> ### Blocking on Flow Control {#blocking}

Author: I'm merging this subsection into the data limits increment section

reset. STOP_SENDING is most useful for streams in the "Recv" or "Size Known"
states.

An endpoint is expected to send another STOP_SENDING frame if a packet
containing a previous STOP_SENDING is lost.  However, once either all stream
data or a RST_STREAM frame has been received for the stream - that is, the
stream is in any state other than "Recv" or "Size Known" - sending a
STOP_SENDING frame is unnecessary.

# Flow Control {#flow-control}

It is necessary to limit the amount of data that a sender may have outstanding
at any time, so as to prevent a fast sender from overwhelming a slow receiver,
or to prevent a malicious sender from consuming significant resources at a
receiver.  To this end, QUIC employs a credit-based flow-control scheme similar
to that in HTTP/2 {{?HTTP2}}.  A receiver advertises the number of octets it is
prepared to receive on a given stream and for the entire connection.  This leads
to two levels of flow control in QUIC:

* Stream flow control, which prevents a single stream from consuming the entire
  receive buffer for a connection.

* Connection flow control, which prevents senders from exceeding a receiver's
  buffer capacity for the connection, and

> A data receiver sends MAX_STREAM_DATA or MAX_DATA frames to the sender to

Reviewer 3: I think we should add a paragraph here that goes like this:

"Endpoints SHOULD communicate initial values for stream and connection flow control using transport parameters {#transport-params}. Data senders MUST use a value of zero for limits not advertised in the peer's transport parameters."

Reviewer 2: The paragraph above looks good enough to me.  I don't think we need to set defaults (if this were the only thing that is read, a reader would (mostly correctly) infer that the values are always set.

STREAM frames on all streams.  A receiver advertises credit for a connection by
sending a MAX_DATA frame.  A receiver maintains a cumulative sum of bytes
received on all contributing streams, which are used to check for flow control
violations. A receiver might use a sum of bytes consumed on all streams to
determine the maximum data limit to be advertised.

A receiver MAY advertise a larger offset at any point by sending MAX_STREAM_DATA
or MAX_DATA frames.  A receiver cannot renege on an advertisement; that is, once
a receiver advertises an offset, advertising a smaller offset has no effect.  A
sender MUST therefore ignore any MAX_STREAM_DATA or MAX_DATA frames that do not
increase flow control limits.

A receiver MUST close the connection with a FLOW_CONTROL_ERROR error
({{error-handling}}) if the peer violates the advertised connection or stream
data limits.

A sender SHOULD send STREAM_BLOCKED or BLOCKED frames to indicate it has data to
write but is blocked by flow control limits.  These frames are expected to be
sent infrequently in common cases, but they are considered useful for debugging
and monitoring purposes.

> (TODO: Add something about max_stream_id)

Reviewer 2: ```suggestion
A similar method is used to control the number of open streams
see {{stream-limit-increment}} for details.
```

connection level flow control.  Given enough time, both endpoints must agree on
flow control state.  If one end believes it can send more than the other end is
willing to receive, the connection will be torn down when too much data arrives.
Conversely if a sender believes it is blocked, while endpoint B expects more
data can be received, then the connection can be in a deadlock, with the sender
waiting for a MAX_STREAM_DATA or MAX_DATA frame which will never come.

On receipt of a RST_STREAM frame, an endpoint will tear down state for the
matching stream and ignore further data arriving on that stream.  This could
result in the endpoints getting out of sync, since the RST_STREAM frame may have
arrived out of order and there may be further bytes in flight.  The data sender
would have counted the data against its connection level flow control budget,
but a receiver that has not received these bytes would not know to include them
as well.  The receiver must learn the number of bytes that were sent on the
stream to make the same adjustment in its connection flow controller.

To ensure that endpoints maintain a consistent connection-level flow control

> state, the RST_STREAM frame {{frame-rst-stream}} includes the largest offset of

Reviewer 2: ```suggestion
state, the RST_STREAM frame ({{frame-rst-stream}}) includes the largest offset of
```

state for closed streams, which could mean a significant state commitment.

## Flow Control for Cryptographic Handshake {#flow-control-crypto}

Data sent in CRYPTO frames is not flow controlled in the same way as STREAM
frames.  QUIC relies on the cryptographic protocol implementation to avoid
excessive buffering of data, see {{QUIC-TLS}}.  The implementation SHOULD
provide an interface to QUIC to tell it about its buffering limits so that there
is not excessive buffering at multiple layers.

## Stream Limit Increment {#stream-limit-increment}

An endpoint limits the number of concurrently active incoming streams by
limiting the maximum stream ID.  An initial value is set in the transport
parameters (see {{transport-parameter-definitions}}) and is subsequently
increased by MAX_STREAM_ID frames (see {{frame-max-stream-id}}).

As with stream and connection flow control, this document leaves when and how
many streams to make available to a peer via MAX_STREAM_ID to implementations,
but offers a few considerations.  MAX_STREAM_ID frames constitute minimal
overhead, while withholding MAX_STREAM_ID frames can prevent the peer from using
the available parallelism.

> 

Reviewer 2: ```suggestion

The STREAM_ID_BLOCKED frame ({{frame-stream-id-blocked}}) can be
used to signal a shortage of available streams.
```

Implementations will likely want to increase the maximum stream ID as
peer-initiated streams close.  A receiver MAY also advance the maximum stream ID
based on current activity, system conditions, and other environmental factors.

# Connections {#connections}

A QUIC connection is a single conversation between two QUIC endpoints.  QUIC's

> connection establishment intertwines version negotiation with the cryptographic

Reviewer 2: ```suggestion
connection establishment combines version negotiation with the cryptographic
```

and transport handshakes to reduce connection establishment latency, as
described in {{handshake}}.  Once established, a connection may migrate to a

> different IP or port at either endpoint, due to NAT rebinding or mobility, as

Reviewer 2: ```suggestion
different IP or port at either endpoint as
```

described in {{migration}}.  Finally, a connection may be terminated by either
endpoint, as described in {{termination}}.

## Connection ID {#connection-id}

> Each connection possesses a set of identifiers, any of which could be used to

Reviewer 2: ```suggestion
Each connection possesses a set of connection identifiers, or connection IDs,
each of which can be identify the connection.  Connection IDs are independently
selected by endpoints; each endpoint selects the connection IDs that its peer
uses.
```

Multi-line suggestion, sorry, I don't know how to hit the entire paragraph.

> distinguish it from other connections.  Connection IDs are selected

Reviewer 2: ```suggestion
```

> independently in each direction.

Reviewer 2: ```suggestion
```

this means the same connection ID MUST NOT be issued more than once on the same
connection.

A zero-length connection ID MAY be used when the connection ID is not needed for
routing and the address/port tuple of packets is sufficient to identify a
connection. An endpoint whose peer has selected a zero-length connection ID MUST
continue to use a zero-length connection ID for the lifetime of the connection
and MUST NOT send packets from any other local address.

When an endpoint has requested a non-zero-length connection ID, it needs to
ensure that the peer has a supply of connection IDs from which to choose for
packets sent to the endpoint.  These connection IDs are supplied by the endpoint
using the NEW_CONNECTION_ID frame ({{frame-new-connection-id}}).

### Issuing Connection IDs

Each Connection ID has an associated sequence number to assist in deduplicating
messages.  The initial connection ID issued by an endpoint is sent in the Source
Connection ID field of the long packet header ({{long-header}}) during the
handshake.  The sequence number of the initial connection ID is 0.  If the
preferred_address transport parameter is sent, the sequence number of the

> supplied connection ID is 1. Subsequent connection IDs are communicated to the

Reviewer 2: ```suggestion
supplied connection ID is 1.

Additional connection IDs are communicated to the
```

> peer using NEW_CONNECTION_ID frames ({{frame-new-connection-id}}), and the

Reviewer 2: ```suggestion
peer using NEW_CONNECTION_ID frames ({{frame-new-connection-id}}).  The
```

packets sent from only one local address.  An endpoint that migrates away from a
local address SHOULD retire all connection IDs used on that address once it no
longer plans to use that address.

## Matching Packets to Connections {#packet-handling}

Incoming packets are classified on receipt.  Packets can either be associated
with an existing connection, or - for servers - potentially create a new
connection.

Hosts try to associate a packet with an existing connection. If the packet has a
Destination Connection ID corresponding to an existing connection, QUIC
processes that packet accordingly. Note that more than one connection ID can be
associated with a connection; see {{connection-id}}.

If the Destination Connection ID is zero length and the packet matches the
address/port tuple of a connection where the host did not require connection
IDs, QUIC processes the packet as part of that connection. Endpoints MUST drop
packets with zero-length Destination Connection ID fields if they do not
correspond to a single connection.

Endpoints SHOULD send a Stateless Reset ({{stateless-reset}}) for any packets
that cannot be attributed to an existing connection.

> <!-- TODO: We are missing a paragraph on failed packet protection removal

Reviewer 2: ```suggestion
Packets that are matched to an existing connection, but for which the endpoint
cannot remove packet protection, are discarded.
```

> here. We probably want that.-->

Reviewer 2: ```suggestion
```

### Client Packet Handling {#client-pkt-handling}

Valid packets sent to clients always include a Destination Connection ID that

> matches the value the client selects.  Clients that choose to receive

Reviewer 2: ```suggestion
matches a value the client selects.  Clients that choose to receive
```

frame carrying the ClientHello.

Note that multiple QUIC packets -- even of different encryption levels -- may be
coalesced into a single UDP datagram (see {{packet-coalesce}}), and so this
handshake may consist of as few as 4 UDP datagrams, or any number more. For
instance, the server's first flight contains packets from the Initial encryption
level (obfuscation), the Handshake level, and "0.5-RTT data" from the server at
the 1-RTT encryption level.

~~~~
Client                                                  Server

Initial[0]: CRYPTO[CH] ->

                                 Initial[0]: CRYPTO[SH] ACK[0]
                       Handshake[0]: CRYPTO[EE, CERT, CV, FIN]
                                 <- 1-RTT[0]: STREAM[1, "..."]

Initial[1]: ACK[0]
Handshake[0]: CRYPTO[FIN], ACK[0]
1-RTT[0]: STREAM[0, "..."], ACK[0] ->

                           1-RTT[1]: STREAM[55, "..."], ACK[0]
                                       <- Handshake[1]: ACK[0]
~~~~
{: #tls-1rtt-handshake title="Example 1-RTT Handshake"}

{{tls-0rtt-handshake}} shows an example of a connection with a 0-RTT handshake
and a single packet of 0-RTT data. Note that as described in {{packet-numbers}},

> the server ACKs the 0-RTT data at the 1-RTT encryption level, and the client's

Reviewer 2: ```suggestion
the server acknowledges 0-RTT data at the 1-RTT encryption level, and the
client sends 1-RTT packets in the same packet number space.
```

> sequence numbers at the 1-RTT encryption level continue to increment from its

Reviewer 2: ```suggestion
```

> 0-RTT packets.

Reviewer 2: ```suggestion
```

defined in this document is possible, the Frame Type field for these frames is
encoded on a single octet.  For instance, though 0x4007 is a legitimate
two-octet encoding for a variable-length integer with a value of 7, PING frames
are always encoded as a single octet with the value 0x07.  An endpoint MUST
treat the receipt of a frame type that uses a longer encoding than necessary as
a connection error of type PROTOCOL_VIOLATION.

# Packetization and Reliability {#packetization}

A sender bundles one or more frames in a QUIC packet (see {{frames}}).

A sender SHOULD minimize per-packet bandwidth and computational costs by
bundling as many frames as possible within a QUIC packet.  A sender MAY wait for
a short period of time to bundle multiple frames before sending a packet that is
not maximally packed, to avoid sending out large numbers of small packets.  An
implementation may use knowledge about application sending behavior or
heuristics to determine whether and for how long to wait.  This waiting period
is an implementation decision, and an implementation should be careful to delay
conservatively, since any delay is likely to increase application-visible
latency.

> 

Reviewer 2: This can be moved to the address validation section.

> 

Reviewer 2: We covered this in packet number spaces already.  Maybe it can go.

> 

Reviewer 2: Move to section on packet number sizes.  (I think that it can be deleted now).

> 

Reviewer 2: Thus far, we have not really addressed 0-RTT at all.  We don't even really say much about what it is for.   This really belongs in a section for that, probably right after the example handshake exchanges.

> 

Reviewer 2: This is old text, I think.  It doesn't even mention Initial, which we now need before any Handshake packets can be sent.

> 

Reviewer 2: Move this and the next paragraph to address validation section.

We might also need a xref from the packet size section to that section as well.

> 

Reviewer 2: We use this encoding in packet headers, so this really needs to be earlier, as I said.

> 

Reviewer 2: Error codes and versions.

| 0x10 - 0x17 | STREAM               | {{frame-stream}}               |
| 0x18        | CRYPTO               | {{frame-crypto}}               |
| 0x19        | NEW_TOKEN            | {{frame-new-token}}            |
| 0x1a - 0x1b | ACK                  | {{frame-ack}}                  |
{: #frame-types title="Frame Types"}

All QUIC frames are idempotent.  That is, a valid frame does not cause
undesirable side effects or errors when received more than once.

The Frame Type field uses a variable length integer encoding (see
{{integer-encoding}}) with one exception.  To ensure simple and efficient
implementations of frame parsing, a frame type MUST use the shortest possible
encoding.  Though a two-, four- or eight-octet encoding of the frame types
defined in this document is possible, the Frame Type field for these frames is
encoded on a single octet.  For instance, though 0x4007 is a legitimate
two-octet encoding for a variable-length integer with a value of 7, PING frames
are always encoded as a single octet with the value 0x07.  An endpoint MUST
treat the receipt of a frame type that uses a longer encoding than necessary as
a connection error of type PROTOCOL_VIOLATION.

> 

Reviewer 2: ```suggestion
  {{QUIC-RECOVERY}}, until either all data has been acknowledged.
```

> 

Reviewer 2: ```suggestion
```

> 

Reviewer 2: ```suggestion
```

I think that this is fine as it is.  If it were longer, maybe, but it's a nice size now and the two concepts are very closely related.

> 

Reviewer 2: ```suggestion

<!-- TODO: delete this section after confirming that it is redundant -->
```

> 

Reviewer 2: ```suggestion

<!-- TODO: split out definition of PreferredAddress -->

```

> 

Reviewer 2: I'm not sure that I like the fact that the specific transport parameters are described way back up top now.  I think that having the description of each down here is better.  There's a lot of encoding information in them, and we don't generally rely on the definitions of transport parameters for much.  Anything we do need can use a forward reference.

> 

Reviewer 2: ```suggestion
```

> 

Reviewer 2: ```suggestion
```

> 

Reviewer 2: ```suggestion
```

> 

Reviewer 2: ```suggestion
```

> 

Reviewer 2: ```suggestion
An endpoint can use the sequence number supplied in the NEW_CONNECTION_ID frame
to identify connection IDs that it has already received.
```

  connection error of type PROTOCOL_VIOLATION.

Sequence Number:

: The sequence number assigned to the connection ID by the sender.  See
  {{issuing-connection-ids}}.

Connection ID:

: A connection ID of the specified length.

Stateless Reset Token:

: A 128-bit value that will be used for a stateless reset when the associated
  connection ID is used (see {{stateless-reset}}).

An endpoint MUST NOT send this frame if it currently requires that its peer send
packets with a zero-length Destination Connection ID.  Changing the length of a
connection ID to or from zero-length makes it difficult to identify when the
value of the connection ID changed.  An endpoint that is sending packets with a
zero-length Destination Connection ID MUST treat receipt of a NEW_CONNECTION_ID
frame as a connection error of type PROTOCOL_VIOLATION.

Transmission errors, timeouts and retransmissions might cause the same
NEW_CONNECTION_ID frame to be received multiple times.  Receipt of the same
frame multiple times MUST NOT be treated as a connection error.  A receiver can
use the sequence number supplied in the NEW_CONNECTION_ID frame to identify new

> connection IDs from old ones.

Reviewer 2: I think that we can remove the first two sentences of this paragraph.  They are vestiges of the design prior to us putting sequence numbers back in.

Each ACK Block acknowledges a contiguous range of packets by indicating the
number of acknowledged packets that precede the largest packet number in that
block.  A value of zero indicates that only the largest packet number is
acknowledged.  Larger ACK Block values indicate a larger range, with
corresponding lower values for the smallest packet number in the range.  Thus,
given a largest packet number for the ACK, the smallest value is determined by
the formula:

~~~
   smallest = largest - ack_block
~~~

The range of packets that are acknowledged by the ACK Block include the range
from the smallest packet number to the largest, inclusive.

The largest value for the First ACK Block is determined by the Largest
Acknowledged field; the largest for Additional ACK Blocks is determined by
cumulatively subtracting the size of all preceding ACK Blocks and Gaps.

Each Gap indicates a range of packets that are not being acknowledged.  The
number of packets in the gap is one higher than the encoded value of the Gap
Field.

The value of the Gap field establishes the largest packet number value for the
ACK Block that follows the gap using the following formula:

> ~~~

Reviewer 2: We should move the remainder of this section to the acknowledgments section.

> ~~~

Reviewer 2: Same here.

> ~~~

Reviewer 2: This paragraph and the next probably belong in the packetization section.
